File: clients/python/gateway_client.py
"""
Order Gateway Client — Python (async)

Drop this into 100-ALGO to replace direct Kite API calls.
Provides: send_order(), with idempotency key + retry on gateway failure.

Usage:
    from gateway_client import GatewayClient

    client = GatewayClient(
        gateway_url=settings.ORDER_GATEWAY_URL,
        api_key=settings.ORDER_GATEWAY_API_KEY,
    )
    result = await client.send_order(
        exchange="NFO",
        tradingsymbol="NIFTY24MAR24100CE",
        transaction_type="SELL",
        quantity=50,
        product="MIS",
        order_type="MARKET",
        source="100-ALGO",
    )
"""

import asyncio
import uuid
import logging
from dataclasses import dataclass, field
from typing import Optional, Literal
import httpx

logger = logging.getLogger(__name__)

Exchange = Literal["NSE", "BSE", "NFO", "BFO", "MCX", "CDS"]
TransactionType = Literal["BUY", "SELL"]
Product = Literal["CNC", "MIS", "NRML"]
OrderType = Literal["MARKET", "LIMIT", "SL", "SL-M"]
Variety = Literal["regular", "amo", "co", "iceberg"]


@dataclass
class OrderResult:
    success: bool
    order_id: Optional[str] = None
    message: Optional[str] = None
    latency_ms: int = 0


@dataclass
class GatewayClient:
    """
    Async HTTP client for the Order Gateway.

    Creates a persistent httpx.AsyncClient with keep-alive connections
    for minimal overhead on repeated calls.
    """

    gateway_url: str
    api_key: str
    timeout_ms: int = 8000
    max_retries: int = 2
    retry_delay_ms: int = 300
    source: str = "100-ALGO"

    _client: Optional[httpx.AsyncClient] = field(default=None, init=False, repr=False)

    def __post_init__(self) -> None:
        self.gateway_url = self.gateway_url.rstrip("/")

    def _get_client(self) -> httpx.AsyncClient:
        """Lazy-initialize persistent client with keep-alive."""
        if self._client is None or self._client.is_closed:
            self._client = httpx.AsyncClient(
                base_url=self.gateway_url,
                headers={
                    "X-API-Key": self.api_key,
                    "Content-Type": "application/json",
                    "Connection": "keep-alive",
                },
                timeout=httpx.Timeout(self.timeout_ms / 1000),
                # Keep-alive pool: persist connections between requests
                limits=httpx.Limits(max_keepalive_connections=5, max_connections=10),
            )
        return self._client

    async def send_order(
        self,
        exchange: Exchange,
        tradingsymbol: str,
        transaction_type: TransactionType,
        quantity: int,
        product: Product,
        order_type: OrderType,
        *,
        variety: Variety = "regular",
        price: Optional[float] = None,
        trigger_price: Optional[float] = None,
        tag: Optional[str] = None,
        idempotency_key: Optional[str] = None,
    ) -> OrderResult:
        """
        Send an order to the gateway.

        The idempotency_key is auto-generated if not provided.
        The same key can be sent multiple times safely — the gateway
        will return the cached result after the first execution.
        """
        payload: dict = {
            "idempotencyKey": idempotency_key or str(uuid.uuid4()),
            "source": self.source,
            "exchange": exchange,
            "tradingsymbol": tradingsymbol,
            "transactionType": transaction_type,
            "quantity": quantity,
            "product": product,
            "orderType": order_type,
            "variety": variety,
        }
        if price is not None:
            payload["price"] = price
        if trigger_price is not None:
            payload["triggerPrice"] = trigger_price
        if tag is not None:
            payload["tag"] = tag

        return await self._post_with_retry("/order", payload, self.max_retries)

    async def health(self) -> dict:
        """Check gateway health (no auth required)."""
        try:
            client = self._get_client()
            res = await client.get("/health")
            return res.json()
        except Exception as e:
            return {"status": "unreachable", "error": str(e)}

    async def _post_with_retry(
        self, path: str, payload: dict, retries_left: int
    ) -> OrderResult:
        try:
            client = self._get_client()
            res = await client.post(path, json=payload)
            data = res.json()

            # Client-side errors (400, 409) — don't retry
            if res.status_code < 500:
                return OrderResult(
                    success=data.get("success", False),
                    order_id=data.get("orderId"),
                    message=data.get("message"),
                    latency_ms=data.get("latencyMs", 0),
                )

            # 5xx — retry if we have attempts left
            if retries_left > 0:
                logger.warning(
                    "Gateway returned %d, retrying (%d left)",
                    res.status_code,
                    retries_left,
                )
                await asyncio.sleep(self.retry_delay_ms / 1000)
                return await self._post_with_retry(path, payload, retries_left - 1)

            return OrderResult(
                success=False,
                message=f"Gateway error: HTTP {res.status_code}",
                latency_ms=data.get("latencyMs", 0),
            )

        except (httpx.TimeoutException, httpx.ConnectError, httpx.NetworkError) as e:
            if retries_left > 0:
                logger.warning(
                    "Gateway network error: %s, retrying (%d left)", e, retries_left
                )
                await asyncio.sleep(self.retry_delay_ms / 1000)
                # Re-create client on network error
                await self.close()
                return await self._post_with_retry(path, payload, retries_left - 1)

            logger.error("Gateway unreachable after retries: %s", e)
            return OrderResult(success=False, message=f"Gateway unreachable: {e}")

    async def close(self) -> None:
        """Close the HTTP client. Call on app shutdown."""
        if self._client and not self._client.is_closed:
            await self._client.aclose()
            self._client = None

    async def __aenter__(self) -> "GatewayClient":
        return self

    async def __aexit__(self, *_) -> None:
        await self.close()

File: clients/node/gateway-client.ts
/**
 * Order Gateway Client — Node.js/TypeScript
 *
 * Drop this file into ultra-order to replace direct Kite API calls.
 * Provides: sendOrder(), with built-in idempotency + retry on gateway failure.
 *
 * Usage:
 *   import { gatewayClient } from './gateway-client';
 *   const result = await gatewayClient.sendOrder({ ... });
 */

import { randomUUID } from 'crypto';

export interface GatewayOrderRequest {
  /** Unique key to prevent duplicate orders. Generate with crypto.randomUUID() */
  idempotencyKey?: string;
  /** Which app is placing this order (shows in dashboard) */
  source: string;
  exchange: 'NSE' | 'BSE' | 'NFO' | 'BFO' | 'MCX' | 'CDS';
  tradingsymbol: string;
  transactionType: 'BUY' | 'SELL';
  quantity: number;
  product: 'CNC' | 'MIS' | 'NRML';
  orderType: 'MARKET' | 'LIMIT' | 'SL' | 'SL-M';
  variety?: 'regular' | 'amo' | 'co' | 'iceberg';
  price?: number;
  triggerPrice?: number;
  tag?: string;
}

export interface GatewayOrderResponse {
  success: boolean;
  orderId?: string;
  message?: string;
  latencyMs: number;
}

export interface GatewayClientOptions {
  /** Gateway base URL e.g. http://gateway-ip:3000 */
  gatewayUrl: string;
  /** API key (GATEWAY_API_KEY in gateway's .env) */
  apiKey: string;
  /** Request timeout in ms. Default: 8000 */
  timeoutMs?: number;
  /** Max retries if gateway is unreachable. Default: 2 */
  maxRetries?: number;
  /** Delay between retries in ms. Default: 300 */
  retryDelayMs?: number;
}

export class GatewayClient {
  private readonly url: string;
  private readonly apiKey: string;
  private readonly timeoutMs: number;
  private readonly maxRetries: number;
  private readonly retryDelayMs: number;

  constructor(opts: GatewayClientOptions) {
    this.url = opts.gatewayUrl.replace(/\/$/, '');
    this.apiKey = opts.apiKey;
    this.timeoutMs = opts.timeoutMs ?? 8000;
    this.maxRetries = opts.maxRetries ?? 2;
    this.retryDelayMs = opts.retryDelayMs ?? 300;
  }

  async sendOrder(req: GatewayOrderRequest): Promise<GatewayOrderResponse> {
    const payload = {
      ...req,
      idempotencyKey: req.idempotencyKey ?? randomUUID(),
      variety: req.variety ?? 'regular',
    };

    return this.postWithRetry('/order', payload, this.maxRetries);
  }

  async health(): Promise<{ status: string; kiteConnected: boolean }> {
    const res = await fetch(`${this.url}/health`, {
      signal: AbortSignal.timeout(3000),
    });
    return res.json() as Promise<{ status: string; kiteConnected: boolean }>;
  }

  private async postWithRetry(
    path: string,
    body: unknown,
    retriesLeft: number
  ): Promise<GatewayOrderResponse> {
    try {
      const res = await fetch(`${this.url}${path}`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          'X-API-Key': this.apiKey,
          Connection: 'keep-alive',
        },
        body: JSON.stringify(body),
        signal: AbortSignal.timeout(this.timeoutMs),
      });

      const data = await res.json() as GatewayOrderResponse;

      // Non-5xx failures (validation errors, duplicate key) — don't retry
      if (!res.ok && res.status < 500) {
        return data;
      }

      if (!res.ok && retriesLeft > 0) {
        await sleep(this.retryDelayMs);
        return this.postWithRetry(path, body, retriesLeft - 1);
      }

      return data;
    } catch (err: unknown) {
      // Network / timeout errors — retry
      if (retriesLeft > 0) {
        await sleep(this.retryDelayMs);
        return this.postWithRetry(path, body, retriesLeft - 1);
      }
      return {
        success: false,
        message: `Gateway unreachable: ${String(err)}`,
        latencyMs: 0,
      };
    }
  }
}

function sleep(ms: number): Promise<void> {
  return new Promise((r) => setTimeout(r, ms));
}

// ─── Singleton factory ────────────────────────────────────────────────────────
// Configure once from environment, use everywhere

let _client: GatewayClient | null = null;

export function createGatewayClient(): GatewayClient {
  if (_client) return _client;

  const gatewayUrl = process.env.ORDER_GATEWAY_URL;
  const apiKey = process.env.ORDER_GATEWAY_API_KEY;

  if (!gatewayUrl) throw new Error('ORDER_GATEWAY_URL environment variable is required');
  if (!apiKey) throw new Error('ORDER_GATEWAY_API_KEY environment variable is required');

  _client = new GatewayClient({ gatewayUrl, apiKey });
  return _client;
}

export const gatewayClient = {
  sendOrder: (req: GatewayOrderRequest) => createGatewayClient().sendOrder(req),
  health: () => createGatewayClient().health(),
};
